### PR TITLE
User configurable autohide delay

### PR DESCRIPTION
**Pull request by pelatx**
Hi,

This PR adds an entry in View menu which opens a modal dialog with a range slider.

That slider adjust the delay of the outline to be exposed, when we hover over the placeholder.

I supposed a range between 0 and 1000 milliseconds is enougth.

### CODE REVIEW

**Comment by pelatx** [comment]
Ups! Travis-ci wants me to work more.

Tomorrow I'll take a look.

File: src/Autohide.js [hunk 1]
```diff
@@ -182,10 +186,51 @@ define(function (require, exports, module) {
...
+        var Dialog = Dialogs.showModalDialog(
+            brackets.DIALOG_ID_SAVE_CLOSE,
+            Strings.AUTOHIDE_DELAY_MODAL_TITLE,
+            Mustache.render(sliderTemplate, {
+                exposeDelay: exposeDelay + Strings.AUTOHIDE_DELAY_MODAL_MS_LABEL
+            }),
+            [{
+                className: Dialogs.DIALOG_BTN_CLASS_PRIMARY,
+                id: "outline.cancel",
+                text: Strings.AUTOHIDE_DELAY_MODAL_BUTTON_CANCEL
```
Comment (Hirse) [inline src/Autohide.js]: In order to reduce the number of translations and stay closer to Bracket's UX, I would go with the standard dialog and only have one Ok-button:
http://brackets.io/docs/current/modules/widgets/Dialogs.html#-showModalDialog

That way, we also get Close-On-Esc for free (which could be seen as cancel).

File: templates/delaySlider.html [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<form action="#">
+    <p class="outline-range-field">
+        <input type="range" id="outline-delay-slider" min="0" max="1000" value="{{exposeDelay}}" />
+    </p>
+    <p id="current-selected-delay">{{exposeDelay}}</p>
```
Comment (Hirse) [inline templates/delaySlider.html]: Could you make the current value also an input field for those that prefer to type the value?

File: styles/styles.css [hunk 3]
```diff
@@ -122,6 +122,37 @@
...
+    -webkit-appearance: none;
+    margin: 18px 0;
+    width: 100%;
+}
+
+#outline-delay-slider:focus {
+  outline: none;
+}
+
+#outline-delay-slider::-webkit-slider-runnable-track {
```
Comment (Hirse) [inline styles/styles.css]: I like the styling that you have done, but it somehow doesn't really look in-place to me.

Would you mind playing around with color and size a bit?
Maybe try the handle in the button color (with a similar focus indicator) and the bar in a dark grey?

File: nls/root/strings.js [hunk 4]
```diff
@@ -11,7 +11,8 @@ define({
     COMMAND_UNNAMED: "Show Unnamed Functions",
     COMMAND_ARGS: "Show Arguments",
     COMMAND_INDENT: "Indent Entries",
-    COMMAND_AUTOHIDE: "Outline List Autohide",
+    COMMAND_AUTOHIDE: "Outline List (Enable Autohide)",
```
Comment (Hirse) [inline nls/root/strings.js]: I don't particularly like this renaming (and naming scheme). None of the other menu entries have parens.

I would maybe go with `Autohide Outline List` and `Set Outline List Autohide Delay`.

File: src/Autohide.js [hunk 5]
```diff
@@ -182,10 +186,51 @@ define(function (require, exports, module) {
         exposeDelay = delay;
     }
 
+    function showDelayDialog() {
+        var Dialog = Dialogs.showModalDialog(
+            brackets.DIALOG_ID_SAVE_CLOSE,
+            Strings.AUTOHIDE_DELAY_MODAL_TITLE,
+            Mustache.render(sliderTemplate, {
+                exposeDelay: exposeDelay + Strings.AUTOHIDE_DELAY_MODAL_MS_LABEL
```
Comment (Hirse) [inline src/Autohide.js]: This doesn't seem to work for setting the current slider value.

**Comment by Hirse** [comment]
Regarding the csslint errors, I think two of them should be disabled:
1. `Using height with border can sometimes make elements larger than you expect.` (maybe warning)
2. `The property -moz-appearance is compatible with -webkit-appearance and should be included as well.` (Brackets is Chrome shell only, right? ... then again, it doesn't hurt to include it)

File: src/Autohide.js (hunk 1 above)
Comment (pelatx) [reply src/Autohide.js]: Fixed.

File: styles/styles.css (hunk 3 above)
Comment (pelatx) [reply styles/styles.css]: I will play a bit.

File: nls/root/strings.js (hunk 4 above)
Comment (pelatx) [reply nls/root/strings.js]: Fixed.

File: src/Autohide.js (hunk 5 above)
Comment (pelatx) [reply src/Autohide.js]: Ups! I wanted to go too fast . It worked until I added the string ... hehe

File: templates/delaySlider.html (hunk 2 above)
Comment (pelatx) [reply templates/delaySlider.html]: I'll try :wink:

File: styles/styles.css (hunk 3 above)
Comment (pelatx) [reply styles/styles.css]: Something like this:

outline-autohide-delay-dark
outline-autohide-delay-light

**Comment by pelatx** [comment]
I'm going to commit the changes to check if it passes the tests and so you can find something else that needs to be changed.

Even if the comments were corrected, we have to see if the style of the slider is correct.

File: nls/root/strings.js [hunk 6]
```diff
@@ -30,5 +31,8 @@ define({
     PREF_AUTOHIDE_NAME: "Outline - Autohide",
     PREF_AUTOHIDE_DESC: "True to enable autohide of the Outline List",
     PREF_AUTOHIDE_DELAY_NAME: "Outline - Autohide Delay",
-    PREF_AUTOHIDE_DELAY_DESC: "Milliseconds of delay for the Outline List to be exposed when the mouse is over the placeholder"
+    PREF_AUTOHIDE_DELAY_DESC: "Milliseconds of delay for the Outline List to be exposed when the mouse is over the placeholder",
+
+    AUTOHIDE_DELAY_MODAL_TITLE: "Outline List - Autohide Expose Delay",
+    AUTOHIDE_DELAY_MODAL_MS_LABEL: " milliseconds"
```
Comment (Hirse) [inline nls/root/strings.js]: Please format the HTML so we don't have to rely on the leading space in the translation strings.

File: nls/root/strings.js [hunk 7]
```diff
@@ -30,5 +31,8 @@ define({
     PREF_AUTOHIDE_NAME: "Outline - Autohide",
     PREF_AUTOHIDE_DESC: "True to enable autohide of the Outline List",
     PREF_AUTOHIDE_DELAY_NAME: "Outline - Autohide Delay",
-    PREF_AUTOHIDE_DELAY_DESC: "Milliseconds of delay for the Outline List to be exposed when the mouse is over the placeholder"
+    PREF_AUTOHIDE_DELAY_DESC: "Milliseconds of delay for the Outline List to be exposed when the mouse is over the placeholder",
```
Comment (Hirse) [inline nls/root/strings.js]: Can we re-use this string in the Dialog to show some explanation?

File: src/Autohide.js [hunk 8]
```diff
@@ -3,10 +3,14 @@ define(function (require, exports, module) {
...
     /* beautify preserve:start *//* eslint-disable no-multi-spaces */
     var Mustache            = brackets.getModule("thirdparty/mustache/mustache");
+    var Dialogs             = brackets.getModule("widgets/Dialogs");
 
     var OutlineManager      = require("src/OutlineManager");
     var prefs               = require("src/Preferences");
+    var Strings             = require("strings");
+
     var placeholderTemplate = require("text!templates/placeholder.html");
+    var sliderTemplate      = require("text!templates/delaySlider.html");
```
Comment (Hirse) [inline src/Autohide.js]: Please rename the file to `delayDialog` or something similar (I would like to have `Dialog` in the file name).

File: src/Autohide.js [hunk 9]
```diff
@@ -182,10 +186,55 @@ define(function (require, exports, module) {
...
+                millisecondsLabel: Strings.AUTOHIDE_DELAY_MODAL_MS_LABEL
+            })
+        );
+
+        var $slider = $("#outline-delay-slider");
+        var $inputbox = $("#outline-delay-inputbox");
+        var inputboxCurrentValue = $inputbox.val();
+
+        Dialog.done(function (buttonId) {
+            if (buttonId === "ok") {
```
Comment (Hirse) [inline src/Autohide.js]: Please use the [Dialog Buttons IDs](http://brackets.io/docs/current/modules/widgets/Dialogs.html#-DIALOG_BTN_CANCEL) constants.

File: src/Autohide.js [hunk 10]
```diff
@@ -182,10 +186,55 @@ define(function (require, exports, module) {
...
+            Strings.AUTOHIDE_DELAY_MODAL_TITLE,
+            Mustache.render(sliderTemplate, {
+                exposeDelay: exposeDelay,
+                millisecondsLabel: Strings.AUTOHIDE_DELAY_MODAL_MS_LABEL
+            })
+        );
+
+        var $slider = $("#outline-delay-slider");
+        var $inputbox = $("#outline-delay-inputbox");
+        var inputboxCurrentValue = $inputbox.val();
```
Comment (Hirse) [inline src/Autohide.js]: Please rename to just `currentValue` because it represents both, the input box and the slider.

Comment (Hirse) [inline src/Autohide.js]:
```suggestion
        var inputboxCurrentValue = exposeDelay;
```

Please keep the value as int to prevent casting.

File: src/Autohide.js [hunk 11]
```diff
@@ -182,10 +186,55 @@ define(function (require, exports, module) {
...
+
+        Dialog.done(function (buttonId) {
+            if (buttonId === "ok") {
+                prefs.set("autohideDelay", $slider.val());
```
Comment (Hirse) [inline src/Autohide.js]: This sets the preference as string. While that still works, I would prefer it to be set as int.

File: src/Autohide.js [hunk 12]
```diff
@@ -182,10 +186,55 @@ define(function (require, exports, module) {
...
+
+        // Slider change handler
+        $slider.on("input", function () {
+            inputboxCurrentValue = $slider.val();
+            $inputbox.val(inputboxCurrentValue);
+        });
+
+        // Inputbox handler
+        $inputbox.keyup(function (ev) {
+            if ((ev.keyCode < 48 || ev.keyCode > 57) && ev.keyCode !== 8) {
```
Comment (Hirse) [inline src/Autohide.js]: Are there any keyCode constants that can be used here?
If not, please add comments so a reader doesn't have to look up the codes.

File: src/Autohide.js [hunk 13]
```diff
@@ -182,10 +186,55 @@ define(function (require, exports, module) {
...
+                $(this).val(parseInt(inputboxCurrentValue, 10));
+            } else if ($(this).val() < 0 || $(this).val() > 1000) {
+                $(this).val(parseInt(inputboxCurrentValue, 10));
+            } else {
+                if ($(this).val() === "") {
+                    $(this).val(0);
```
Comment (Hirse) [inline src/Autohide.js]: I'm not sure about this behavior. It feels a bit irritating. Imagine the workflow:
Current value `300`, want to change to `500`. Focus on input, delete everything and want to type `500`, but the input is never empty. It still works (due to parsing ints), but it doesn't look like it.

Instead, maybe leave the input empty, use `placeholder="0"` in the HTML and treat empty as 0 when parsing.

File: styles/styles.css [hunk 14]
```diff
@@ -4,9 +4,9 @@
     font-weight: normal;
     src: url("fonts/ionicons.eot?v=2.0.0");
     src: url("fonts/ionicons.eot?v=2.0.0#iefix") format("embedded-opentype"),
-         url("fonts/ionicons.ttf?v=2.0.0") format("truetype"),
-         url("fonts/ionicons.woff?v=2.0.0") format("woff"),
-         url("fonts/ionicons.svg?v=2.0.0#Ionicons") format("svg");
+        url("fonts/ionicons.ttf?v=2.0.0") format("truetype"),
+        url("fonts/ionicons.woff?v=2.0.0") format("woff"),
+        url("fonts/ionicons.svg?v=2.0.0#Ionicons") format("svg");
```
Comment (Hirse) [inline styles/styles.css]: Please revert this change.

File: styles/styles.css [hunk 15]
```diff
@@ -122,6 +122,41 @@
...
+    border-radius: 3px;
+    box-shadow: 1px 1px 1px #000000, 0 0 1px #0d0d0d;
+    cursor: pointer;
+    height: 22px;
+    margin-top: -8px;
+    width: 10px;
+}
+
+#outline-delay-slider:focus {
+  outline: none;
```
Comment (Hirse) [inline styles/styles.css]: Please add some focus border/outline to the handle/thumb.
Brackets Button style for comparison (normal and focus):

File: styles/styles.css [hunk 16]
```diff
@@ -122,6 +122,41 @@
...
+    width: 100%;
+}
+
+#outline-delay-slider::-webkit-slider-thumb {
+    -moz-appearance: none;
+    -webkit-appearance: none;
+    background: #368ac7;
+    border-radius: 3px;
+    box-shadow: 1px 1px 1px #000000, 0 0 1px #0d0d0d;
+    cursor: pointer;
```
Comment (Hirse) [inline styles/styles.css]: Maybe the `grab` cursor would be more appropriate here.

File: styles/styles.css (hunk 3 above)
Comment (Hirse) [reply styles/styles.css]: Yes, I like this much better. Thanks.

File: styles/styles.css [hunk 17]
```diff
@@ -122,6 +122,41 @@
     transition: right 0.2s ease-out;
 }
 
+#outline-delay-slider {
```
Comment (Hirse) [inline styles/styles.css]:
```suggestion

/* ==== Outline Autohide Delay Dialog ==== */

#outline-delay-slider {
```

File: nls/root/strings.js (hunk 6 above)
Comment (pelatx) [reply nls/root/strings.js]: Sorry for the delay. I've been busy the last few days.

Fixed.

File: src/Autohide.js (hunk 8 above)
Comment (pelatx) [reply src/Autohide.js]: Fixed.

File: src/Autohide.js (hunk 9 above)
Comment (pelatx) [reply src/Autohide.js]: Fixed.

File: src/Autohide.js (hunk 10 above)
Comment (pelatx) [reply src/Autohide.js]: Fixed.

Comment (pelatx) [reply src/Autohide.js]: Ups! fixed.

File: src/Autohide.js (hunk 13 above)
Comment (pelatx) [reply src/Autohide.js]: Take a look at this and let me know what you think, please.
```
        // Inputbox handler on focus
        $inputbox.focus(function () {
            $(this).select();
        });

        // Inputbox handler to prevent not numeric or backspace input
        $inputbox.keypress(function (ev) {
            // The numbers (0-9) on the keyboard correspond to this range of key codes (48-57).
            // Key code 8 correspond to backspace key.
            // Only these eleven keys are allowed.
            if ((ev.keyCode < 48 || ev.keyCode > 57) && ev.keyCode !== 8) {
                ev.preventDefault();
            }
        });

        // Input handler to ensure a valid value in any case
        $inputbox.keyup(function () {
            var parsed = parseInt($(this).val(), 10);
            if (!$.isNumeric(parsed)) {
                parsed = 0;
            }
            if (parsed > 1000) {
                currentValue = 1000;
                $(this).val(currentValue);
            } else {
                currentValue = parsed;
            }
            $slider.val(currentValue);
        });
```

File: src/Autohide.js (hunk 11 above)
Comment (pelatx) [reply src/Autohide.js]: Fixed.

File: styles/styles.css (hunk 14 above)
Comment (pelatx) [reply styles/styles.css]: Sorry. I use automatic indentation and I did not realize. Fixed.

File: nls/root/strings.js (hunk 7 above)
Comment (Hirse) [reply nls/root/strings.js]: Yes, as first element in the "body" of the dialog sounds good.

File: src/Autohide.js (hunk 13 above)
Comment (Hirse) [reply src/Autohide.js]: Yes, I think that looks better.

Can the key validation be done by using [`type="number"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number) on the input?

File: styles/styles.css (hunk 15 above)
Comment (pelatx) [reply styles/styles.css]: I have not been able to style the handle without doing it with the track as well. And I really dislike that outline on the whole slider when it's focused.

I have removed the entry for focus and I can not do much more. Sorry.

File: styles/styles.css (hunk 16 above)
Comment (pelatx) [reply styles/styles.css]: Fixed.
